test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc

```diff
@@ -531,25 +552,31 @@ TEST_P(SslTapIntegrationTest, RequestWithTextProto) {
   EXPECT_FALSE(trace.socket_buffered_trace().write_truncated());
 }
 
-// Validate a single request with JSON (body as string) output.
-TEST_P(SslTapIntegrationTest, RequestWithJsonBodyAsString) {
-  max_rx_bytes_ = 4;
-  max_tx_bytes_ = 5;
+// Validate a single request with JSON (body as string) output. This test uses an upstream tap.
+TEST_P(SslTapIntegrationTest, RequestWithJsonBodyAsStringUpstreamTap) {
+  upstream_tap_ = true;
+  max_rx_bytes_ = 5;
+  max_tx_bytes_ = 4;
 
   format_ = envoy::service::tap::v2alpha::OutputSink::JSON_BODY_AS_STRING;
   ConnectionCreationFunction creator = [&]() -> Network::ClientConnectionPtr {
     return makeSslClientConnection({});
   };
-  const uint64_t id = Network::ConnectionImpl::nextGlobalIdForTest() + 1;
+  const uint64_t id = Network::ConnectionImpl::nextGlobalIdForTest() + 2;
   testRouterRequestAndResponseWithBody(512, 1024, false, &creator);
   checkStats();
   codec_client_->close();
   test_server_->waitForCounterGe("http.config_test.downstream_cx_destroy", 1);
+  test_server_.reset();
+
+  // This must be done after server shutdown so that connection pool connections are closed and
+  // the tap written.
   envoy::data::tap::v2alpha::TraceWrapper trace;
   MessageUtil::loadFromFile(fmt::format("{}_{}.json", path_prefix_, id), trace, *api_);
+
   // Test some obvious properties.
-  EXPECT_EQ(trace.socket_buffered_trace().events(0).read().data().as_string(), "POST");
-  EXPECT_EQ(trace.socket_buffered_trace().events(1).write().data().as_string(), "HTTP/");
+  EXPECT_EQ(trace.socket_buffered_trace().events(0).write().data().as_string(), "POST");
+  EXPECT_EQ(trace.socket_buffered_trace().events(1).read().data().as_string(), "HTTP/");
```

Comment:
It seems like you modified a previous test to do something different. TBH I'm not fully understanding the semantic difference. But I do wonder whether it might have made sense to keep the old function and add a new one. Was the old functionality of the RequestWithJsonBodyAsString test not needed?

Reply:
The difference is the tap filter is installed on the upstream connection vs. the downstream. This test was primarily testing a different output type, but now we cover two different things. I can bring back the old test but I'm not sure it adds much extra value.